File: dotnet/examples/Wrangling/SessionTrees/PutAndRemoveBranchMappingTable.cs
/**
 * Copyright © 2024 Diffusion Data Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

using System;
using System.Threading;
using System.Threading.Tasks;
using PushTechnology.ClientInterface.Client.Factories;
using static System.Console;
using static PushTechnology.ClientInterface.Examples.Program;

namespace PushTechnology.ClientInterface.Examples.Wrangling.SessionTrees
{
    public sealed class PutAndRemoveBranchMappingTable : Example
    {
        public override async Task Run(CancellationToken cancellationToken, string[] args)
        {
            string serverUrl = args[0];

            var session = Diffusion.Sessions
                .Principal("admin")
                .Credentials(Diffusion.Credentials.Password("password"))
                .Open(serverUrl);

            var table = Diffusion.NewBranchMappingTableBuilder()
                .AddBranchMapping("$Principal is 'admin'", "my/topic/path/for/admin")
                .AddBranchMapping("$Principal is 'control'", "my/topic/path/for/control")
                .AddBranchMapping("$Principal is ''", "my/topic/path/for/anonymous")
                .Create("my/personal/path");

            await session.SessionTrees.PutBranchMappingTableAsync(table, cancellationToken);

            WriteLine($"Session tree mappings added.");

            await session.SessionTrees.PutBranchMappingTableAsync(Diffusion.NewBranchMappingTableBuilder().Create("my/personal/path"), cancellationToken);

            WriteLine($"Session tree mapping table and mappings removed.");

            session.Close();
        }
    }
}

File: python/examples/src/diffusion_examples/wrangling/session_trees/put_and_remove_branch_mapping_table.py
# Copyright © 2024 Diffusion Data Ltd.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import asyncio
from diffusion import sessions, Credentials
from diffusion_examples.utils.program import Example
from diffusion.features.control.session_trees import BranchMappingTable


class PutAndRemoveBranchMappingTable(Example):
    async def run(
        self,
        server_url: str = "<url>",
        principal: str = "<principal>",
        password: str = "<password>",
    ):
        async with sessions().principal(principal).credentials(Credentials(password)).open(
            server_url
        ) as session:
            table = (
                BranchMappingTable.Builder()
                .add_branch_mapping("$Principal is 'admin'", "my/topic/path/for/admin")
                .add_branch_mapping("$Principal is 'control'", "my/topic/path/for/control")
                .add_branch_mapping("$Principal is ''", "my/topic/path/for/anonymous")
                .create("my/personal/path")
            )

            await session.session_trees.put_branch_mapping_table(table)
            print("Session tree mappings added.")
            await session.session_trees.put_branch_mapping_table(
                BranchMappingTable.Builder().create("my/personal/path")
            )
            print("Session tree mapping table and mappings removed.")



if __name__ == '__main__':
    asyncio.run(PutAndRemoveBranchMappingTable().run(
        server_url="ws://localhost:8080",
        principal="admin",
        password="password",
    ))
